pkg/chunk/elastic/es_index_client.go

```diff
@@ -0,0 +1,316 @@
+}
+
+var client *elastic.Client
+
+// Starting with elastic.v5, you must pass a context to execute each service
+var ctx = context.Background()
+
+// StorageClient implements chunk.IndexClient and chunk.ObjectClient for Cassandra.
+type esClient struct {
+	cfg    Config
+	client *elastic.Client
+}
+
+func (e *esClient) Stop() {
+	e.client.Stop()
+}
+
+// WriteBatch will be transferred into a bulk request to fulfill batch write rather than parallel write
+type writeBatch struct {
+	entries []chunk.IndexEntry
+}
+
+func (e *esClient) NewWriteBatch() chunk.WriteBatch {
+	return &writeBatch{}
+}
+
+func (b *writeBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	b.entries = append(b.entries, chunk.IndexEntry{
+		TableName:  tableName,
+		HashValue:  hashValue,
+		RangeValue: rangeValue,
+		Value:      value,
+	})
+}
+
+func (e *esClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch) error {
+	b := batch.(*writeBatch)
+
+	indexName := b.entries[0].TableName
+
+	bulkRequest := e.client.Bulk()
+	for _, entry := range b.entries {
+		index := IndexEntry{Hash: entry.HashValue, Range: string(entry.RangeValue), Value: string(entry.Value)}
+		req := elastic.NewBulkIndexRequest().Index(indexName).Type(e.cfg.IndexType).Doc(index)
+		bulkRequest = bulkRequest.Add(req)
+	}
+
+	err := instrument.CollectedRequest(ctx, "ES.BatchWrite", esRequestDuration,
+		instrument.ErrorCode, func(ctx context.Context) error {
+			var err error
+			_, err = bulkRequest.Do(ctx)
+			return err
+		})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// readBatch represents a batch of rows read from ElasticSearch.
+type readBatch struct {
+	rangeValue []byte
+	value      []byte
+}
+
+func (r readBatch) Iterator() chunk.ReadBatchIterator {
+	return &elasticReadBatchIterator{
+		readBatch: r,
+	}
+}
+
+type elasticReadBatchIterator struct {
+	consumed bool
+	readBatch
+}
+
+func (r *elasticReadBatchIterator) Next() bool {
+	if r.consumed {
+		return false
+	}
+	r.consumed = true
+	return true
+}
+
+func (r *elasticReadBatchIterator) RangeValue() []byte {
+	return r.rangeValue
+}
+
+func (r *elasticReadBatchIterator) Value() []byte {
+	return r.value
+}
+
+func (e *esClient) QueryPages(ctx context.Context, queries []chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error {
+	return chunk_util.DoParallelQueries(ctx, e.query, queries, callback)
+}
+
+func (e *esClient) query(ctx context.Context, query chunk.IndexQuery, callback func(chunk.ReadBatch) (shouldContinue bool)) error {
+	var rangeQuery *elastic.RangeQuery
+	var valueTermQuery *elastic.TermQuery
+
+	level.Debug(util.Logger).Log("msg", fmt.Sprintf(
+		"hash [%s], rangeValuePrefix [%s], rangeValueStart [%s]", query.HashValue, query.RangeValuePrefix, query.RangeValueStart))
+	hashTermQuery := elastic.NewTermQuery("hash", query.HashValue)
+	switch {
+	case len(query.RangeValuePrefix) > 0 && query.ValueEqual == nil:
```

Comment:
Comments for each case would assist others reading this to understand the intent of the code.